library/Freckle/App/Stats.hs

```diff
@@ -111,14 +115,33 @@ withStatsClient StatsSettings {..} f = do
   if amsEnabled
     then do
       tags <- (amsTags <>) <$> getEcsMetadataTags
-
+      gauges <- liftIO $ newIORef mempty
       Datadog.withDogStatsD amsSettings $ \client ->
         -- Add the tags to the thread context so they're present in all logs
         withThreadContext (map toPair tags)
-          $ f StatsClient { scClient = client, scTags = tags }
-    else f $ StatsClient { scClient = Datadog.Dummy, scTags = amsTags }
+          $ f StatsClient { scClient = client, scTags = tags, scGauges = gauges }
+    else do
+      gauges <- liftIO $ newIORef mempty
+      f $ StatsClient { scClient = Datadog.Dummy, scTags = amsTags, scGauges = gauges }
   where toPair = bimap (fromString . unpack) String
 
+withGauge :: (MonadReader app m, HasStatsClient app, MonadUnliftIO m) => Text -> m a -> m a
+withGauge name f = do
+  gaugesRef <- asks $ scGauges . view statsClientL
+  gaugesMap <- liftIO $ readIORef gaugesRef
+  gaugeValue <- liftIO $ maybe (createNewEKG gaugesRef gaugesMap) pure (HashMap.lookup name gaugesMap)
+  bracketEKG f gaugeValue
+  where
+    createNewEKG ref hashmap = do
+      ekg <- EKG.new
+      liftIO $ writeIORef ref (HashMap.insert name ekg hashmap)
+      pure ekg
+    bracketEKG action value = do
+      liftIO $ value `EKG.add` 1
+      result <- action
+      liftIO $ value `EKG.subtract` 1
+      pure result
```

Comment:
You're calling this "bracket" but you're not actually bracketing it; in other words, an exception in `action` won't decrement the gauge. I think you want something like,

```suggestion
    bracketEKG action value = do
      liftIO $ value `EKG.add` 1
      action `finally` liftIO (EKG.subtract 1)
```

Reply:
Ah, I think I meant to actually use bracket and just spaced on it.